examples/fabric-admin/commands/pairing/DeviceSubscription.cpp

```diff
@@ -0,0 +1,168 @@
+{
+    reinterpret_cast<DeviceSubscription *>(context)->OnDeviceConnected(exchangeMgr, sessionHandle);
+}
+
+void OnDeviceConnectionFailureWrapper(void * context, const ScopedNodeId & peerId, CHIP_ERROR error)
+{
+    reinterpret_cast<DeviceSubscription *>(context)->OnDeviceConnectionFailure(peerId, error);
+}
+
+} // namespace
+
+DeviceSubscription::DeviceSubscription() :
+    mOnDeviceConnectedCallback(OnDeviceConnectedWrapper, this),
+    mOnDeviceConnectionFailureCallback(OnDeviceConnectionFailureWrapper, this)
+{}
+
+void DeviceSubscription::OnAttributeData(const ConcreteDataAttributePath & path, TLV::TLVReader * data, const StatusIB & status)
+{
+    VerifyOrDie(path.mEndpointId == kRootEndpointId);
+    VerifyOrDie(path.mClusterId == Clusters::AdministratorCommissioning::Id);
+
+    switch (path.mAttributeId)
+    {
+    case Clusters::AdministratorCommissioning::Attributes::WindowStatus::Id: {
+        Clusters::AdministratorCommissioning::CommissioningWindowStatusEnum windowStatus;
+        CHIP_ERROR err = data->Get(windowStatus);
+        VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(NotSpecified, "Failed to read WindowStatus"));
+        VerifyOrReturn(windowStatus != Clusters::AdministratorCommissioning::CommissioningWindowStatusEnum::kUnknownEnumValue);
+        mCurrentAdministratorCommissioningAttributes.window_status = static_cast<uint32_t>(windowStatus);
+        mChangeDetected                                            = true;
+        break;
+    }
+    case Clusters::AdministratorCommissioning::Attributes::AdminFabricIndex::Id: {
+        FabricIndex fabricIndex;
+        CHIP_ERROR err                                                       = data->Get(fabricIndex);
+        mCurrentAdministratorCommissioningAttributes.has_opener_fabric_index = err == CHIP_NO_ERROR;
+        if (mCurrentAdministratorCommissioningAttributes.has_opener_fabric_index)
+        {
+            mCurrentAdministratorCommissioningAttributes.opener_fabric_index = static_cast<uint32_t>(fabricIndex);
+        }
+        mChangeDetected = true;
+        break;
+    }
+    case Clusters::AdministratorCommissioning::Attributes::AdminVendorId::Id: {
+        chip::VendorId vendorId;
+        CHIP_ERROR err                                                    = data->Get(vendorId);
+        mCurrentAdministratorCommissioningAttributes.has_opener_vendor_id = err == CHIP_NO_ERROR;
+        if (mCurrentAdministratorCommissioningAttributes.has_opener_vendor_id)
+        {
+            mCurrentAdministratorCommissioningAttributes.opener_vendor_id = static_cast<uint32_t>(vendorId);
+        }
+        mChangeDetected = true;
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void DeviceSubscription::OnReportEnd()
+{
+    // Report end is at the end of all attributes (success)
+    if (mChangeDetected)
+    {
+#if defined(PW_RPC_ENABLED)
+        // TODO we need to schedule work instead of calling this directly
```

Comment:
Why need to reschedule, OnReportEnd is not called from the CHIP main thread?

Reply:
What thread is is called on?

Reply:
But this would still block a packet received call stack on RPC call when we can process further packets if that is needed first.

I still think there is an advantage to scheduling the work. If you don't think so I will remove TODO and just carry on

Reply:
RPC is running in its own thread, if you do schedule here, it just schedule the work back to the same thread which run  DeviceSubscription::OnReportEnd.

Reply:
Yes, it will block on RPC call, if you want DeviceSubscription::OnReportEnd earlier, you could schedule to your-self.